File: fp-in-ocaml/exercises/state/rng.ml
(** 乱数生成器のモジュール型 *)
module type RNG = sig
  type t
  (** 乱数生成器自体の型。 抽象型のため、外部から実態を参照することは出来ない。 *)

  type 'a rand = t -> 'a * t
  (** 乱数生成のための型。 *)

  val make : int64 -> t
  (** [int64] から乱数生成器を作る *)

  val next_int : t -> int * t
  (** 乱数生成器を基にして、整数の乱数を生成する。 *)
end

module Simple : RNG = struct
  type t = int64
  type 'a rand = t -> 'a * t

  let make = Fun.id

  let next_int seed =
    (* [Int64.t]のための演算子がないので shadowing で置き換える *)
    let ( + ) = Int64.add in
    let ( * ) = Int64.mul in

    (* [land],[lsr]などは文字列だがニ項演算子として使える *)
    let ( land ) = Int64.logand in
    let ( lsr ) = Int64.shift_right_logical in

    let seed = ((seed * 0x5DEECE66DL) + 0xBL) land 0xFFFFFFFFFFFFL in

    (seed lsr 16 |> Int64.to_int, seed)
end

(** ファンクターと呼ばれる機能により、RNGからいくつかの関数を導出する。

    圏論の文脈における関手のファンクターとは全く別物であることに注意。

    ファンクターはモジュールを引数にとり、モジュールを生成する。
    標準ライブラリなどでも、ファンクターによるモジュールの生成は[Make]という名前であることが多い。

    @see <https://ocaml.org/docs/functors> Functors *)
module Make (Rng : RNG) = struct
  include Rng

  let int : int rand = Rng.next_int
  let unit (a : 'a) : 'a rand = fun rng -> (a, rng)

  let map (f : 'a -> 'b) (s : 'a rand) : 'b rand =
   fun rng -> s rng |> Pair.map_fst f

  (** Exercise 6.1: 非負整数をランダム生成する関数[non_negative_int]を実装せよ。 *)
  let non_negative_int (_rng : t) : int * t = failwith "Not implemented"

  (** Exercise 6.2: 0以上1未満の浮動小数点数をランダム生成する関数[double]を実装せよ。 *)
  let double (_rng : t) : float * t = failwith "Not implemented"

  (** Exercise 6.3: 整数と浮動小数点数の組を生成する[int_double], [double_int]、
      浮動小数点数の3つ組を生成する[double3]を実装せよ。 *)

  let int_double (_rng : t) : (int * float) * t = failwith "Not implemented"
  let double_int (_rng : t) : (float * int) * t = failwith "Not implemented"

  let double3 (_rng : t) : (float * float * float) * t =
    failwith "Not implemented"

  (** Exercise 6.4: 引数[count]で指定された要素数の整数リストを生成する[ints]を実装せよ。 *)
  let ints (_count : int) (_rng : t) : int list * t = failwith "Not implemented"

  (** Exercise 6.5: [map]を用いて[double]を再実装せよ。 *)
  let double_via_map (_rng : t) : float * t = failwith "Not implemented"

  (** Exercise 6.6: 2つのRand値を関数[f]で合成する[map2]を実装せよ。 *)
  let map2 (_f : 'a -> 'b -> 'c) (_ra : 'a rand) (_rb : 'b rand) : 'c rand =
    failwith "Not implemented"

  (** Exercise 6.7: Rand値のリストをまとめる[sequence]を実装せよ。 *)
  let sequence (_rs : 'a rand list) : 'a list rand = failwith "Not implemented"

  (** Exercise 6.8: 1つ前の生成結果に依存して次のRand値を構成する[flat_map]を実装せよ。 *)
  let flat_map (_f : 'a -> 'b rand) (_r : 'a rand) : 'b rand =
    failwith "Not implemented"

  (** Exercise 6.9: [flat_map]を用いて[map], [map2]を再実装せよ。 *)

  let map_via_flat_map (_f : 'a -> 'b) (_r : 'a rand) : 'b rand =
    failwith "Not implemented"

  let map2_via_flat_map (_f : 'a -> 'b -> 'c) (_ra : 'a rand) (_rb : 'b rand) :
      'c rand =
    failwith "Not implemented"
end

File: fp-in-haskell/src/FpInHaskell/Answers/State/RNG.hs
module FpInHaskell.Answers.State.RNG (
    RNG (..),
    nextInt,
    Rand,
    int,
    unit,
    map,
    nonNegativeInt,
    double,
    boolean,
    intDouble,
    doubleInt,
    double3,
    ints,
    doubleViaMap,
    map2,
    sequence,
    flatMap,
    nonNegativeLessThan,
    mapViaFlatMap,
    map2ViaFlatMap,
) where

import Data.Bits (shiftR, (.&.))
import Data.Int (Int32)
import Data.Word (Word64)
import Prelude hiding (map, sequence)

-- Prelude には同名の `map`/`sequence` がすでにあるため、上の `import Prelude hiding (...)` で
-- それらを読み込み対象から外し、ここでは `Rand` に対する同名の演習関数を自前で定義している。

-- 乱数生成器。この演習では単一の実装(線形合同法)しか使わないため、型クラスや抽象データ型で
-- 一般化せず、具体的な `newtype` として素直に定義する。
newtype RNG = Simple Word64
    deriving (Show)

-- Javaの `java.util.Random` と同じ線形合同法(LCG)。48bit のシード値を使う。
-- `newSeed` は `Word64` 上で48bitマスクして計算するが、返す乱数 `n` は32bit符号付き整数
-- (`Int32`)として解釈しなければならない(32bit目が立っていれば負の値になる)。
-- そのため一度 `Int32` を経由してから `Int` へ拡張する。
nextInt :: RNG -> (Int, RNG)
nextInt (Simple seed) = (n, Simple newSeed)
  where
    newSeed = (seed * 0x5DEECE66D + 0xB) .&. 0xFFFFFFFFFFFF
    n = fromIntegral (fromIntegral (newSeed `shiftR` 16) :: Int32)

-- `RNG` を状態として持ち回す計算を表す型。
type Rand a = RNG -> (a, RNG)

int :: Rand Int
int = nextInt

unit :: a -> Rand a
unit x rng = (x, rng)

-- Prelude の `map` と同じ引数順(関数、Rand の順)。
map :: (a -> b) -> Rand a -> Rand b
map f s rng = let (x, rng2) = s rng in (f x, rng2)

-- Exercise 6.1: 非負整数をランダム生成する関数 `nonNegativeInt` を実装せよ。
--
-- `minBound :: Int` の絶対値は `maxBound :: Int` より1大きいため、負の数に1を足してから
-- 正の数にする。これにより `minBound` は `maxBound` に、`-1` は `0` に写る。
nonNegativeInt :: RNG -> (Int, RNG)
nonNegativeInt rng =
    let (i, r) = nextInt rng
     in (if i < 0 then -(i + 1) else i, r)

-- Exercise 6.2: 0以上1未満の浮動小数点数をランダム生成する関数 `double` を実装せよ。
double :: RNG -> (Double, RNG)
double rng =
    let (i, r) = nonNegativeInt rng
     in (fromIntegral i / (fromIntegral (maxBound :: Int) + 1), r)

boolean :: RNG -> (Bool, RNG)
boolean rng = let (i, rng2) = nextInt rng in (even i, rng2)

-- Exercise 6.3: 整数と浮動小数点数の組をランダム生成する関数 `intDouble` と `doubleInt` を実装せよ。
-- また、浮動小数点数の3つ組をランダム生成する関数 `double3` を実装せよ。
intDouble :: RNG -> ((Int, Double), RNG)
intDouble rng =
    let (i, r1) = nextInt rng
        (d, r2) = double r1
     in ((i, d), r2)

doubleInt :: RNG -> ((Double, Int), RNG)
doubleInt rng =
    let ((i, d), r) = intDouble rng
     in ((d, i), r)

double3 :: RNG -> ((Double, Double, Double), RNG)
double3 rng =
    let (d1, r1) = double rng
        (d2, r2) = double r1
        (d3, r3) = double r2
     in ((d1, d2, d3), r3)

-- Exercise 6.4: 引数で指定された要素数の整数リストをランダム生成する関数 `ints` を実装せよ。
ints :: Int -> RNG -> ([Int], RNG)
ints count rng
    | count <= 0 = ([], rng)
    | otherwise =
        let (x, r1) = nextInt rng
            (xs, r2) = ints (count - 1) r1
         in (x : xs, r2)

-- Exercise 6.5: `map` を用いて `double` を実装せよ。
doubleViaMap :: Rand Double
doubleViaMap = map (\i -> fromIntegral i / (fromIntegral (maxBound :: Int) + 1)) nonNegativeInt

-- Exercise 6.6: 関数 `map2` を実装せよ。
map2 :: (a -> b -> c) -> Rand a -> Rand b -> Rand c
map2 f ra rb rng0 =
    let (x, rng1) = ra rng0
        (y, rng2) = rb rng1
     in (f x y, rng2)

-- Exercise 6.7: 関数 `sequence` を実装せよ。
sequence :: [Rand a] -> Rand [a]
sequence = foldr (\r acc -> map2 (:) r acc) (unit [])

-- Exercise 6.8: 関数 `flatMap` を実装せよ。
flatMap :: (a -> Rand b) -> Rand a -> Rand b
flatMap f r rng0 = let (x, rng1) = r rng0 in f x rng1

-- `flatMap` の使用例。演習番号はないが、`flatMap` の使い方の見本として Answers/Exercises
-- 両方で実装しておく。`n` 未満の非負整数を、偏りが出ないように再試行しながら生成する。
nonNegativeLessThan :: Int -> Rand Int
nonNegativeLessThan n = flatMap go nonNegativeInt
  where
    go i =
        let m = i `mod` n
         in if i + (n - 1) - m >= 0 then unit m else nonNegativeLessThan n

-- Exercise 6.9: `flatMap` を用いて `map`, `map2` を実装せよ。
mapViaFlatMap :: (a -> b) -> Rand a -> Rand b
mapViaFlatMap f = flatMap (unit . f)

map2ViaFlatMap :: (a -> b -> c) -> Rand a -> Rand b -> Rand c
map2ViaFlatMap f ra rb = flatMap (\x -> map (f x) rb) ra
